### PR TITLE
Update compute plan

### PR DESCRIPTION
**Pull request by thibaultrobert**
PR #80 should be merged first

### CODE REVIEW

**Comment by jmorel** [comment]
Tested against all the companion PRs, the full suite of test passes! Well done!

File: chaincode/compute_plan_test.go [hunk 1]
```diff
@@ -643,3 +642,39 @@ func checkComputePlanMetrics(t *testing.T, db *LedgerDB, cpID string, doneCount,
...
+				DataSampleKeys: []string{trainDataSampleHash1},
+				AlgoKey:        algoHash,
+				ID:             NewID,
+				InModelsIDs:    []string{traintupleID1, traintupleID2},
+			},
+		},
+	}
+	out, err = updateComputePlanInternal(db, out.ComputePlanID, up)
+	assert.NoError(t, err)
+	assert.Contains(t, out.IDToKey, NewID)
```
Comment (AurelienGasser) [inline chaincode/compute_plan_test.go:678]: nit: add test to validate the length of `out.IDToKey`?

File: chaincode/compute_plan_dag.go [hunk 2]
```diff
@@ -100,6 +104,12 @@ func (dag *ComputeDAG) sort() error {
 	return nil
 }
 
+func (dag *ComputeDAG) addAndSort(new *ComputeDAG) error {
+	new.sort()
+	dag = new
+	return nil
+}
+
```
Comment (AurelienGasser) [inline chaincode/compute_plan_dag.go]: unused

File: chaincode/compute_plan_dag.go [hunk 3]
```diff
@@ -59,25 +61,27 @@ func createComputeDAG(cp inputComputePlan) (ComputeDAG, error) {
...
 	for i := 0; len(current) != 0; {
 		depth := 0
 		ready := true
 		for _, ID := range current[i].InModelsIDs {
 			if ID == "" {
 				continue
 			}
-			parentDepth, ok := IDPresents[ID]
+			parent, ok := dag.IDtoItem[ID]
 			ready = ready && ok
```
Comment (AurelienGasser) [inline chaincode/compute_plan_dag.go:75]: small optimisation:

```suggestion
			ready = ready && ok
			if !ok {
				break
			}
```

File: chaincode/ledger.go [hunk 4]
```diff
@@ -182,15 +182,23 @@ type Testtuple struct {
...
-	Tag                     string    `json:"tag"`
-	TesttupleKeys           []string  `json:"testtupleKeys"`
-	TraintupleKeys          []string  `json:"traintupleKeys"`
-	TupleCount              int       `json:"tupleCount"`
+	AggregatetupleKeys      []string          `json:"aggregatetupleKeys"`
+	AssetType               AssetType         `json:"assetType"`
+	CompositeTraintupleKeys []string          `json:"compositeTraintupleKeys"`
+	DoneCount               int               `json:"DoneCount"`
+	IDToItem                map[string]CPItem `json:"IDToItem"`
+	MaxRank                 int               `json:"maxRank"`
```
Comment (AurelienGasser) [inline chaincode/ledger.go]: unused

**Review by AurelienGasser — DISMISSED** [review]
Looks good! :tada: I added a few minor comments.

Any thoughts on the following renaming, to make it more explicit what "item" refers to?
- `IDToItem` => `IDToTrainTask`
- `CPItem` => `CPTrainTask`
- etc

File: chaincode/compute_plan_dag.go (hunk 2 above)
Comment (thibaultrobert) [reply chaincode/compute_plan_dag.go]: Yes it's the leftover of a different and way too complicated solution I tried before. Will removed it.

File: chaincode/compute_plan_dag.go (hunk 3 above)
Comment (thibaultrobert) [reply chaincode/compute_plan_dag.go:75]: Yep good idea.

File: chaincode/ledger.go (hunk 4 above)
Comment (thibaultrobert) [reply chaincode/ledger.go]: Indeed, let's removed it.

File: chaincode/compute_plan_test.go (hunk 1 above)
Comment (thibaultrobert) [reply chaincode/compute_plan_test.go:678]: Done

**Comment by jmorel** [comment]
Found by @Kelvin-M: you can't update a compute plan with just testtuples.

**Review by Kelvin-M — DISMISSED** [review]

**Review by jmorel — DISMISSED** [review]
Tested against the companion PRs in the backend, cli and tests, everything's working fine! 👍

**Review by AurelienGasser — DISMISSED** [review]

**Review by AurelienGasser — APPROVED** [review]
LGTM

**Pull request merged**